plottr/plot/pyqtgraph/autoplot.py

```diff
@@ -0,0 +1,375 @@
+        else:
+            self.widget = widget
+
+    # re-implementing to get correct type annotation.
+    def __enter__(self) -> "FigureMaker":
+        return self
+
+    def subPlotFromId(self, subPlotId: int) -> PlotBase:
+        """Get SubPlot from ID."""
+        subPlots = self.subPlots[subPlotId].axes
+        assert isinstance(subPlots, list) and len(subPlots) > 0 and \
+               isinstance(subPlots[0], PlotBase)
+        return subPlots[0]
+
+    def makeSubPlots(self, nSubPlots: int) -> List[PlotBase]:
+        """Create empty subplots in the widgets.
+
+        If ``clearWidget`` was not set to ``True`` in the constructor,
+        existing sub plot widgets are cleared, but not deleted and re-created.
+        """
+        plot: PlotBase
+        if self.clearWidget:
+            self.widget.deleteAllPlots()
+            for i in range(nSubPlots):
+                if max(self.dataDimensionsInSubPlot(i).values()) == 1:
+                    plot = Plot(self.widget)
+                    self.widget.addPlot(plot)
+                elif max(self.dataDimensionsInSubPlot(i).values()) == 2:
+                    plot = PlotWithColorbar(self.widget)
+                    self.widget.addPlot(plot)
+        else:
+            self.widget.clearAllPlots()
+
+        return self.widget.subPlots
+
+    def formatSubPlot(self, subPlotId: int) -> None:
+        """Set labels/legends for the given subplot."""
+        if len(self.plotIdsInSubPlot(subPlotId)) == 0:
+            return
+
+        labels = self.subPlotLabels(subPlotId)
+        subPlot = self.subPlotFromId(subPlotId)
+
+        # label the x axis if there's only one x label
+        if isinstance(subPlot, Plot):
+            if len(set(labels[0])) == 1:
+                subPlot.plot.setLabel("bottom", labels[0][0])
+
+        if isinstance(subPlot, PlotWithColorbar):
+            if len(set(labels[0])) == 1:
+                subPlot.plot.setLabel("bottom", labels[0][0])
+
+            if len(set(labels[1])) == 1:
+                subPlot.plot.setLabel('left', labels[1][0])
+
+            if len(set(labels[2])) == 1:
+                subPlot.colorbar.setLabel('left', labels[2][0])
+
+    def plot(self, plotItem: PlotItem) -> None:
+        """Plot the given item."""
+        if plotItem.plotDataType is PlotDataType.unknown:
+            if len(plotItem.data) == 2:
+                plotItem.plotDataType = PlotDataType.scatter1d
+            elif len(plotItem.data) == 3:
+                plotItem.plotDataType = PlotDataType.scatter2d
+
+        if plotItem.plotDataType in [PlotDataType.scatter1d, PlotDataType.line1d]:
+            self._1dPlot(plotItem)
+        elif plotItem.plotDataType == PlotDataType.grid2d:
+            self._colorPlot(plotItem)
+        elif plotItem.plotDataType == PlotDataType.scatter2d:
+            self._scatterPlot2d(plotItem)
+        else:
+            raise NotImplementedError('Cannot plot this data.')
```

Comment:
users won't see this exception, right? plottr will just crash, and only in the log/terminal users would see this, right? perhaps it's fine, but i is it consistent with other failures where plottr doesn't crash but rather for example refuses to plot?

Reply:
i think if you run this from within a plotnode (as the main application does), then the user would indeed not see it, but also it wouldn't crash. the nodes catch all exceptions and print them as debug info by default.

on the other hand, if this gets raised then there's a bug or unforeseen situation -- since this backend is still not widely tested with all sorts of different data i left it in for now.
i think i'd like to leave this in for now as it might help debugging if someone tries to plot something crazy that i hadn't thought about.